src/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/CryptoStream.cs

```diff
@@ -294,47 +295,78 @@ private void CheckReadArguments(byte[] buffer, int offset, int count)
             int numOutputBytes;
 
             // OK, see first if it's a multi-block transform and we can speed up things
-            if (bytesToDeliver > _outputBlockSize)
+            int blocksToProcess = bytesToDeliver / _outputBlockSize;
+
+            if (blocksToProcess > 1 && _transform.CanTransformMultipleBlocks)
             {
-                if (_transform.CanTransformMultipleBlocks)
+                byte[] tempInputBuffer = null;
+                byte[] tempOutputBuffer = null;
+
+                try
                 {
-                    int BlocksToProcess = bytesToDeliver / _outputBlockSize;
-                    int numWholeBlocksInBytes = BlocksToProcess * _inputBlockSize;
-                    byte[] tempInputBuffer = new byte[numWholeBlocksInBytes];
-                    // get first the block already read
-                    Buffer.BlockCopy(_inputBuffer, 0, tempInputBuffer, 0, _inputBufferIndex);
-                    amountRead = _inputBufferIndex;
-                    amountRead += useAsync ?
+                    int numWholeBlocksInBytes = blocksToProcess * _inputBlockSize;
+                    tempInputBuffer = ArrayPool<byte>.Shared.Rent(numWholeBlocksInBytes);
+
+                    amountRead = useAsync ?
                         await _stream.ReadAsync(new Memory<byte>(tempInputBuffer, _inputBufferIndex, numWholeBlocksInBytes - _inputBufferIndex), cancellationToken) :
                         _stream.Read(tempInputBuffer, _inputBufferIndex, numWholeBlocksInBytes - _inputBufferIndex);
 
-                    _inputBufferIndex = 0;
-                    if (amountRead <= _inputBlockSize)
+                    int totalInput = _inputBufferIndex + amountRead;
+
+                    // If there's still less than a block, copy the new data into the hold buffer and move to the slow read.
+                    if (totalInput < _inputBlockSize)
                     {
-                        _inputBuffer = tempInputBuffer;
-                        _inputBufferIndex = amountRead;
-                        goto slow;
+                        Buffer.BlockCopy(tempInputBuffer, _inputBufferIndex, _inputBuffer, _inputBufferIndex, amountRead);
+                        _inputBufferIndex = totalInput;
                     }
-                    // Make amountRead an integral multiple of _InputBlockSize
-                    int numWholeReadBlocksInBytes = (amountRead / _inputBlockSize) * _inputBlockSize;
-                    int numIgnoredBytes = amountRead - numWholeReadBlocksInBytes;
-                    if (numIgnoredBytes != 0)
+                    else
```

Comment:
I tried looking at the CI code coverage run report, and then trying to do code coverage locally for System.Security.Cryptography.Primitives, but it's not working for that assembly.  Any idea why?  It'd be nice to verify we have tests fully covering all of this.

Reply:
Apparently AnyOS + PartialFacade = No PDB.  I nudged things locally and was content with the coverage.

I've also started inquiring as to why that's the case.  (I suspect the check at https://github.com/dotnet/buildtools/blob/8bfc0315aed0f2011e47b378d388ce7ad7298e1f/src/Microsoft.DotNet.Build.Tasks/PackageFiles/partialfacades.task.targets#L43 intended to include AnyOS, but didn't)

Reply:
Thanks.